sql/src/main/java/io/crate/metadata/Signature.java

```diff
@@ -22,137 +22,363 @@
+        private DataType getCommonType(List<DataType> dataTypes, int start) {
+            DataType commonType = null;
+            for (int i = start; i < dataTypes.size(); i++) {
+                DataType dataType = dataTypes.get(i);
+                if (IS_UNDEFINED.test(dataType)) {
+                    continue;
                 }
-                if (!typeEqualsOrUndefined(givenSignatureIt.next(), varArgsIt.next())) {
-                    return false;
+                if (commonType == null) {
+                    commonType = dataType;
+                } else if (!commonType.equals(dataType)) {
+                    return null;
                 }
             }
-            return !(varArgsIt.hasNext() || givenSignatureIt.hasNext());
+            return commonType == null ? DataTypes.UNDEFINED : commonType;
         }
-    }
 
-    private boolean typeEqualsOrUndefined(DataType givenType, DataType expectedType) {
-        return givenType.id() == UndefinedType.ID || expectedType.equals(givenType);
-    }
+        @Override
+        public List<DataType> apply(List<DataType> dataTypes) {
+            ArgMatcher varArgMatcher = matchers[matchers.length - 1];
+            if (dataTypes.size() != matchers.length) {
+                if (varArgs && dataTypes.size() > matchers.length) {
+                    if (checkVarArgTypes) {
+                        DataType commonType = getCommonType(dataTypes, matchers.length - 1);
+                        if (commonType == null) {
+                            return null;
+                        } else if (!IS_UNDEFINED.test(commonType)) {
+                            // enforce the common type in varArgs
+                            varArgMatcher = dt -> commonType;
+                        }
+                    }
+                } else {
+                    return null;
+                }
+            }
+            return normalize(dataTypes, varArgMatcher);
+        }
 
-    @Override
-    public boolean equals(Object o) {
-        if (this == o) return true;
-        if (o == null || !(o instanceof List)) return false;
-        List otherList = (List) o;
-        return list.equals(otherList);
-    }
+        @Nullable
+        private List<DataType> normalize(List<DataType> dataTypes, ArgMatcher varArgMatcher){
+            for (int i = 0; i < dataTypes.size(); i++) {
+                ArgMatcher matcher = i < matchers.length - 1 ? matchers[i] : varArgMatcher;
+                DataType dataType = dataTypes.get(i);
+                DataType repl = matcher.apply(dataType);
+                if (repl == null) {
+                    return null;
+                } else if (repl != dataType) {
+                    return copy(dataTypes, varArgMatcher);
+                }
+            }
+            return dataTypes;
+        }
 
-    @Override
-    protected List<DataType> delegate() {
-        return list;
+        private List<DataType> copy(List<DataType> in, ArgMatcher varArgMatcher) {
```

Comment:
maybe rename copy to `copyAndReplace` or `createTransformedTypes`.
I think `copy` is a bit missleading as it doesn't just copy the list.